### PR TITLE
GlusterFS cherry-pick Release-3.7 E2E

### PR DESCRIPTION
**Pull request by mjudeikis**
Cherry-pick: #7336 #7326 #7320 #7267 #7237  #7163  #7149 #6918 #6938 #6969  #6614 #6182  #5723

Progress:

s - skip due either ouf of scope or not compatible/not needed for 3.6
c - cherry-picked

cc: @jarrpa 
I think this is good to go for the review.
Doing testing now on aws with 3 gluster nodes and 3 registry nodes. Will ping env details privately if you want to check it out.

Still pending https://github.com/openshift/openshift-ansible/pull/7205 

```
[mjudeiki@redhat openshift-ansible]$ git-oldest-ancestor                                                                                                                       
                                                                                                                                                                                                                                                                                                                
7900f45d2d8ef566cf3f330a33364dd217a93c1d                                                                                       

git lg1 --oneline 7900f45d2d8ef566cf3f330a33364dd217a93c1d..master roles/openshift_storage_glusterfs
c  * 762c9b3f5 GlusterFS: Don't copy non-existant topology file
s * 39ad419b8 Refactor openshift.common.version to openshift_current_version
s *   aa7c020f6 Merge pull request #7322 from mjudeikis/gluster-block-iscsi
  |\  
s | * 261139cff (origin/gluster-block-iscsi, gluster-block-iscsi) enable iscsid on start and add rpcbind dependencies
c * |   f8bc212a6 Merge pull request #7336 from SaravanaStorageNetwork/fix_s3_service_to_s3_pod
  |\ \  
s | * | 97f7213c8 Fix the gluster-s3 pod label used in gluster-s3 service.
  | |/  
c * |   47c7799a6 Merge pull request #7326 from SaravanaStorageNetwork/fix_pvc_count_verify
  |\ \  
s | * | 512e539c3 Fix for gluster-s3 pvc check count.
  | |/  
c * |   de7b04f93 Merge pull request #7320 from SaravanaStorageNetwork/fix_s3_image
  |\ \  
s | * | e8f5a9aec Fix rhgs-s3 image name
  | |/  
c * | 212bc4c97 Add s3 and block uninstall sections as well.
  |/  
c *   8c18cb0fc Merge pull request #7267 from SaravanaStorageNetwork/templates_fix
  |\  
s | * 8e1982cf3 Update glusterfs-template:  - Add GB_LOGDIR  - failureThreshold as 50 secs
c * |   3efd30b3b Merge pull request #7237 from jarrpa/glusterfs-kernel-mod-fix
  |\ \  
  | |/  
  |/|   
s | * b7aa27d52 GlusterFS: Check for groups in template file
c * |   5b21e30c3 Merge pull request #7163 from chinacoolhacker/master
  |\ \  
  | |/  
  |/|   
c | * 95bcc815c Fix for support multi-cluster heketi's topology
c * |   a2dfbe8bd Merge pull request #7149 from jarrpa/uninstall-regression
  |\ \  
s | * | 96080146d GlusterFS: Fix uninstall regression
  | |/  
c  * | 4d9075feb (origin/glusterblock-target-mountpoint) Add target mount for gluster block
c * | f7ef6abaf (origin/kernel_module_loading) kernel module loading fix
  |/  
c  *   b1cca0383 Merge pull request #6918 from SaravanaStorageNetwork/uninstall_playbook
  |\  
s | * 2861f28c0 erase data only if variable is set. fix block indentatation
s | * 81033d580 uninstall playbook for GlusterFS
c * | 27c60fa67 Backport changes for glusterfs, heketi, s3 and block templates
c * | bc712160d GlusterFS: Fix block StorageClass heketi route
s * |   706c31038 Merge pull request #6938 from vshn/revert-heketi-image-option
  |\ \  
s | * | f2ea513cd Revert "GlusterFS: Remove image option from heketi command"
c * | | 674b2ff6c GlusterFS: Minor documentation update
  | |/  
  |/|   
s * | 5402ff05c Initial support for 3.10
c * |   26caf813c Merge pull request #6969 from mjudeikis/gluster-dynamic-invetory-bug
  |\ \  
s | * | ac122861b (origin/gluster-dynamic-invetory-bug) dynamic inventory bug when group exists but its empty
c * | | da60c1b58 (origin/restart_fix_atomic) dynamic inventory bug when group exists but its empty
  |/ /  
c * | 2573825c0 (origin/glusterfs-block-storage) add glusterblock support for ansible
  |/  
s *   fa649787c Merge pull request #6614 from mgugino-upstream-stage/plugins-to-lib-utils
  |\  
s | * d3fefc32a Move more plugins to lib_utils
s  * | ff919a3f4 Add missing v3.9 gluster templates
  |/  
s * e24140eed Add gluster 3.9 templates
s * e3cf9edff Remove references to deployment_type
s * e6c159afb Remove openshift.common.{is_atomic|is_containerized}
s * 801779eeb Relocate filter plugins to lib_utils
s * c113074f5 Deprecate using Ansible tests as filters
s *   3c99f16b7 Merge pull request #6182 from jmencak/glusterfs_storageclass-default
  |\  
c | * 89472a43b Support for making glusterfs storage class a default one.
c * | efe272a51 Remove empty openshift_hosted_facts role			      (WE NOT DO THIS YET? )
  |/  
c  * 0e41e5176 Check for openshift attribute before using it during CNS install.
s * 50751e3c5 Include Deprecation: Convert to include_tasks                           (WE NOT DO THIS YET? )
s * 0de559c8f Fix hosted vars
s *   824a7e6da Merge pull request #6325 from sosiouxme/20171201-glusterfs-bug
  |\  
  | * 010243cef storage_glusterfs: fix typo
s * | fbb4e1ca7 retry package operations
  |/  
c  *   aff556902 Merge pull request #5723 from mjudeikis/gluster-registry-backend
  |\  
  | * e893c818a (origin/gluster-registry-backend) Add support for external glusterfs as registry backend
s * |   b7af6f115 Merge pull request #6073 from mjudeikis/fix_glusterfs_restart
  |\ \  
s | * | 8ff6f015e (origin/fix_glusterfs_restart2, origin/fix_glusterfs_restart) Add check for gluterFS DS   to stop restarts
  | |/  
s * | 9d5bb8d5e GlusterFS: Remove extraneous line from glusterblock template
s * | 7ce0ab6cf GlusterFS: Files and templates for 3.8
  |/  
s * 7ea30979a GlusterFS: Add configuration for auto creating block-hosting volumes
s * cfe73d1eb GlusterFS: Add gluster-s3 functionality
s * 5b85d56f5 GlusterFS: Add glusterblock functionality
s * 80a663769 GlusterFS: Update heketi templates for latest version
s * 9b05d632c GlusterFS: Specify resource requests


git lg1 --oneline 7900f45d2d8ef566cf3f330a33364dd217a93c1d..master roles/openshift_hosted/tasks/registry.yml
[mjudeiki@redhat openshift-ansible]$ git lg1 --oneline 7900f45d2d8ef566cf3f330a33364dd217a93c1d..master roles/openshift_hosted/tasks/registry.yml
s * 195ff0913 Add wait_for_pods to upgrade for hosted components
s * 5ea67bc7b (upstream/revert-6913-revert-6830-registry_addr) Revert "Revert "use non-deprecated REGISTRY_OPENSHIFT_SERVER_ADDR variable to set the registry hostname""
s * b58726426 (upstream/revert-6830-registry_addr) Revert "use non-deprecated REGISTRY_OPENSHIFT_SERVER_ADDR variable to set the registry hostname"
s * 0da4565f5 use non-deprecated REGISTRY_OPENSHIFT_SERVER_ADDR variable to set the registry hostname
s * 28b254886 Move wait_for_pods to it's own play openshift_hosted
s * 26fd036e2 Fix openshift hosted registry rollout
s * 50751e3c5 Include Deprecation: Convert to include_tasks
s * 0de559c8f Fix hosted vars
s * 5120f8e90 Implement container runtime role
s * e893c818a (origin/gluster-registry-backend) Add support for external glusterfs as registry backend (duplicate)
s * 4abb7381b Allow cluster IP for docker-registry service to be set

```

### CODE REVIEW

**Comment by mjudeikis** [comment]
@jarrpa I just did some testing and it looks fine on 3.7. The only thing is that I took "cherry commits, not PR's". For this it might be harder to review :/ sorry

Let me know of any comments, if you want we can jump on call for this and go through together

**Comment by jarrpa** [comment]
@mjudeikis Well, mechanically you did the thing I asked. However, I don't like the fact that you got rid of the names of the original commits. This will not do. Could you please go through and modify the subject lines of all the commits to the subject of their originals? Though do keep the hash of the original commit in the commit message. I'd like it even better if you would get the commit message generated by `git cherry-pick -x` to also preserve the body of the original commit along with the cherry-pick line, but even I'd be hesitant to do that much work.

Also, I can't quite tell offhand: did you include any relevant commits outside of `roles/openshift_storage_glusterfs/`? I believe there were a few in `roles/openshift_hosted/tasks/registry` and then one to modify `roles/openshift_node/templates/multipath.conf.j2`.

**Comment by mjudeikis** [comment]
@jarrpa, ok, will see what can be done :) 
on the files, good spot. Will revisit all this :)

**Comment by DanyC97** [comment]
@mjudeikis you need any help with this? will be awesome if this can be merged in the next 1-3 days.

**Comment by mjudeikis** [comment]
/retest
@DanyC97 its waiting review from @jarrpa

**Comment by mjudeikis** [comment]
@jarrpa checking. 
+tox error - working on it.

**Comment by mjudeikis** [comment]
@jarrpa removed empty one. But we have tox errors. Looks like they are part of release-3.7 branch, as they are failing on HEAD too. Not sure how you want to approach this.

**Comment by DanyC97** [comment]
imo @mjudeikis we should merge it and not gate it.... my 0.002$

**Comment by jarrpa** [comment]
Well, we're not getting the tox errors here yet, so we'll worry about that when we get there. Right now we're still seeing a lot of CI flakes.

I've been waiting on a Trello card to make it through testing before proceeding. It's been sitting still for a week, so I gave it a poke: https://trello.com/c/Hr0GUwDQ/612-2-glusterfs-uninstall-playbook-cns

**Comment by DanyC97** [comment]
@jarrpa your colleague just updated Trello card, does it mean we are good to merge this PR?

**Comment by DanyC97** [comment]
@mjudeikis i am confused tbh where this is being held, you think progress can be made and maybe hope that it can be merged before Easter ?:))

**Comment by DanyC97** [comment]
@jarrpa  any progress on this please ? anything i can help with ?

**Comment by jarrpa** [comment]
@DanyC97 My apologies for the delayed response. I was traveling last week and being pulled in multiple directions two weeks ago, all the while being neck-deep in documentation work for the 3.9 release. :( I have some breathing room now so I'll do a (hopefully) final review today.

File: roles/openshift_storage_glusterfs/templates/v3.6/glusterfs-storageclass.yml.j2 [hunk 1]
```diff
@@ -3,10 +3,6 @@ apiVersion: storage.k8s.io/v1
 kind: StorageClass
 metadata:
   name: glusterfs-{{ glusterfs_name }}
-{% if glusterfs_storageclass_default is defined and glusterfs_storageclass_default %}
-  annotations:
-    storageclass.kubernetes.io/is-default-class: "true"
-{% endif %}
 provisioner: kubernetes.io/glusterfs
```
Comment (jarrpa) [inline roles/openshift_storage_glusterfs/templates/v3.6/glusterfs-storageclass.yml.j2]: Removing this chunk is a bug, but not one introduced by the backport. I'm going to allow this to go forward, submit a PR to fix it against master, and we'll backport it in a different PR. :)

Comment (mjudeikis) [reply roles/openshift_storage_glusterfs/templates/v3.6/glusterfs-storageclass.yml.j2]: cool :) will do

**Comment by jarrpa** [comment]
@sdodson @michaelgugino The Submit Queue check shows green but says "Context retired without replacement". Is that a bad thing, or is this PR actually queued up now?

**Comment by DanyC97** [comment]
@jarrpa no problem i understand ;) Anyone of you ping me if i can help with anything, i just need this in so i can pull a new 3.7 tag with all the fixes .

also @mjudeikis @michaelgugino thx for your help too ;)

**Comment by jarrpa** [comment]
@sdodson @michaelgugino The test failures seem unrelated. Can this be merged?

**Comment by sdodson** [comment]
The install job ran to completion and failed with an e2e test flake. The others are all failures related to CI job configuration on older branches.

**Comment by jarrpa** [comment]
Woo!

@mjudeikis Feel free to start on the 3.6 backport. :) If possible, try not to cherry-pick merge commits. ...although if the PR patchset isn't particularly clean, feel free to use the merge commit.

**Comment by DanyC97** [comment]
w00t, finally i see light at the end of the tunnel :)

thanks all for your help!

**Comment by ahmadou** [comment]
Hello,

On a fresh install of a new cluster, i got this error : 

fatal: [master01.mycompany.com]: FAILED! => {"msg": "The task includes an option with an undefined variable. The error was: 'openshift_client_binary' is undefined\n\nThe error appears to have been in '/home/ansibleuser/openshift-ansible/roles/openshift_storage_glusterfs/tasks/glusterfs_common.yml': line 264, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n- name: Set heketi-cli command\n  ^ here\n"}

The 3.7 install seem to not work for me at all. 
I've been stuck on an error on the glusterfs phase  from heketi pod beeing stuck in crash loop
I thought that this merge was a fix to the glusterfs problem but it seem some variable définition is missing

I don't know if i should open a ticket but the change was introduced by this merge specifically so i took the liberty to comment here.
Sorry if i shouldn't have.

Could you please provide a workaround and /or fix ?

**Comment by mjudeikis** [comment]
@ahmadou 
edit /home/ansibleuser/openshift-ansible/roles/openshift_storage_glusterfs/tasks/glusterfs_common.yml, line 248. 
`openshift_client_binary` -- > `openshift.common.client_binary `

Doing fix now. Slipped with last PR :/

**Comment by mjudeikis** [comment]
fix: https://github.com/openshift/openshift-ansible/pull/7768

**Comment by mjudeikis** [comment]
cc: @DanyC97 , you are impacted too.

**Comment by ahmadou** [comment]
thans for the quick response.

I will apply the workaround

**Comment by mjudeikis** [comment]
Check the PR, there is few more places you need this :)

**Comment by ahmadou** [comment]
I've made the modification.
The install still fails at the glusterfs stage.

I've opened  #7777 (yay jackpot :))

**Comment by mjudeikis** [comment]
so Hekti pod is not starting due unable to mount "DB" volume. Mount from the same server `mount -t glusterfs ip:path /mnt` works, but not from remote server. I bet its iptables rules blocking traffic.

**Comment by mjudeikis** [comment]
@DanyC97 did you did any testing on your end?

**Comment by DanyC97** [comment]
nope cause i'm waiting for a new tag to get out for 3.7 branch. Is easier that way rather then me going again and cherry pick.

And i was told that Tuesday/ Thursday a job is kicked internally but i didn't see that happening.

File: roles/openshift_storage_glusterfs/templates/v3.6/glusterfs-storageclass.yml.j2 (hunk 1 above)
Comment (mjudeikis) [reply roles/openshift_storage_glusterfs/templates/v3.6/glusterfs-storageclass.yml.j2]: @jarrpa what do you want to do with this? this is 3.6 file in 3.7 branch. I made sure this is didn't got into 3.6 branch. And nobody should use 3.6 code in 3.7 branches, so its kinda absolute. Do you want to align them or just leave it as it is.

Comment (jarrpa) [reply roles/openshift_storage_glusterfs/templates/v3.6/glusterfs-storageclass.yml.j2]: Eh, leave it for now. It'll all get deleted when the version collapse PR gets backported.

**Pull request merged**